File: Original/newsloader/src/main/java/com/andruid/magic/newsloader/server/RetrofitClient.java
package com.andruid.magic.newsloader.server;

import android.content.Context;

import com.andruid.magic.newsloader.util.NetworkUtil;

import java.io.IOException;

import okhttp3.Cache;
import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

import static com.andruid.magic.newsloader.data.Constants.BASE_URL;
import static com.andruid.magic.newsloader.data.Constants.CACHE_SIZE;

public class RetrofitClient {
    private static Retrofit retrofit;

    public static Retrofit getRetrofitInstance(final Context context){
        if(retrofit==null) {
            Cache cache = new Cache(context.getCacheDir(), CACHE_SIZE);
            OkHttpClient client = new OkHttpClient.Builder()
                    .cache(cache)
                    .addInterceptor(new Interceptor() {
                        @Override
                        public Response intercept(Chain chain) throws IOException {
                            Request request = chain.request();
                            if(NetworkUtil.hasNetwork(context))
                                request = request.newBuilder().header("Cache-Control",
                                        "public, max-age=" + 5).build();
                            else
                                request = request.newBuilder().header("Cache-Control",
                                        "public, only-if-cached, max-stale=" + 60 * 60 * 24 * 7)
                                        .build();
                            return chain.proceed(request);
                        }
                    })
                    .build();
            retrofit = new Retrofit.Builder()
                    .client(client)
                    .baseUrl(BASE_URL)
                    .addConverterFactory(GsonConverterFactory.create())
                    .build();
        }
        return retrofit;
    }
}

File: NewsDaily/newsloader/src/main/java/com/andruid/magic/newsloader/data/server/RetrofitClient.kt
package com.andruid.magic.newsloader.data.server

import android.content.Context
import com.andruid.magic.newsloader.util.hasNetwork
import okhttp3.Cache
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

object RetrofitClient {
    private const val BASE_URL = "https://news-daily.herokuapp.com"
    private const val CACHE_SIZE = 10L * 1024L * 1024L

    private val LOCK = Any()
    private lateinit var INSTANCE: Retrofit

    fun getRetrofitInstance(context: Context): Retrofit {
        synchronized(LOCK) {
            if (!::INSTANCE.isInitialized) {
                val client = buildClient(context)

                INSTANCE = Retrofit.Builder()
                    .client(client)
                    .baseUrl(BASE_URL)
                    .addConverterFactory(GsonConverterFactory.create())
                    .build()
            }
        }

        return INSTANCE
    }

    private fun buildClient(context: Context): OkHttpClient {
        val cache = Cache(context.cacheDir, CACHE_SIZE)
        return OkHttpClient.Builder()
            .cache(cache)
            .addInterceptor { chain: Interceptor.Chain ->
                var request = chain.request()
                request = if (context.hasNetwork())
                    request.newBuilder().header("Cache-Control", "public, max-age=" + 5)
                        .build()
                else
                    request.newBuilder().header(
                        "Cache-Control",
                        "public, only-if-cached, max-stale=" + 60 * 60 * 24 * 7
                    )
                        .build()
                chain.proceed(request)
            }
            .build()
    }
}